### PR TITLE
Proxmox module refactoring

### PR DESCRIPTION
**Pull request by Lithimlin**
##### SUMMARY

This pull request refactors the historically grown `proxmox` module.
The control flow was very opaque with some redundant checks and - most recently - features that were only introduced in parts of the module when they intended to target the whole module (see #9065).
This refactor aims to make the module more readable and more easily extendable.

Additionally, this request contains various bug fixes:

 - From #9065: The `disk_volume` key is now properly parsed and passed on to the API
 - From #8982: It is now possible to add bind-only mount points again.
 - From #8811: Converting an LXC to a template is now idempotent.
 - From #9055: The default value of `update=False` is now deprecated and will be changed to `update=True` in version 11.0.0
 - In the `proxmox` utils module: The parameter `choose_first_if_multiple` of `get_vmid` is now used correctly, causing the function to return the first of multiple matches instead of failing.

##### ISSUE TYPE

- Bugfix Pull Request
- Refactoring Pull Request

##### COMPONENT NAME

proxmox

##### ADDITIONAL INFORMATION

The following playbook should run correctly on both the old and refactored versions of this module, but the template creation will not be idempotent on the old version:

```yaml
---
- name: Check proxmox LXC module
  hosts: pve
  become: false
  gather_facts: false
  tasks:
    - name: Create LXC
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        ostemplate: local:vztmpl/debian-12-standard_12.2-1_amd64.tar.zst
        disk: local-lvm:8
        tags: test1,test3,debian
      register: create_res

    - name: Sleep for 5 seconds and continue with play
      ansible.builtin.wait_for:
        timeout: 5
      delegate_to: localhost

    - name: Check Creation Idempotency
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        ostemplate: local:vztmpl/debian-12-standard_12.2-1_amd64.tar.zst
        disk: local-lvm:8
        tags: test1,test3,debian
      register: check_res
      failed_when: check_res.changed

    - name: Update LXC config [new - create volumes]
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid | int }}"
        disk_volume:
          storage: local-lvm
          size: 14
        mount_volumes:
          - id: mp0
            storage: local-lvm
            size: 3
            mountpoint: /mnt/test
        update: true

    - name: Update LXC config [new - explicit volumes]
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk_volume:
          storage: local-lvm
          volume: "vm-{{ create_res.vmid }}-disk-0"
          size: 14
        mount_volumes:
          - id: mp0
            storage: local-lvm
            volume: "vm-{{ create_res.vmid }}-disk-1"
            size: 3
            mountpoint: /mnt/test
        update: true

    - name: Update LXC config [old - implicit volumes]
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk: local-lvm:14
        mounts: '{"mp0":"local-lvm:3,mp=/mnt/test"}'
        update: true

    - name: Update LXC config [old - explicit volumes]
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk: "local-lvm:vm-{{ create_res.vmid }}-disk-0,size=14G"
        mounts: '{"mp0":"local-lvm:vm-{{ create_res.vmid }}-disk-1,size=3G,mp=/mnt/test"}'
        update: true

    - name: Update LXC config [tags]
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node
        cores: 1
        vmid: "{{ create_res.vmid }}"
        disk: "local-lvm:vm-{{ create_res.vmid }}-disk-0,size=14G"
        mounts: '{"mp0":"local-lvm:vm-{{ create_res.vmid }}-disk-1,size=3G,mp=/mnt/test"}'
        tags: test1,test2,debian
        update: true

    - name: Start LXC
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: started

    - name: Check Starting Idempotency
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: started
      retries: 1
      delay: 10
      register: check_res
      failed_when: check_res.changed

    - name: Restart LXC
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: restarted

    - name: Stop LXC
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: stopped

    - name: Check Stopping Idempotency
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: stopped
      retries: 1
      delay: 10
      register: check_res
      failed_when: check_res.changed

    - name: Convert LXC to Template
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: template

    - name: Check Template Idempotency
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: template
      retries: 1
      delay: 10
      register: check_res
      failed_when: check_res.changed

    - name: Remove LXC
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: absent

    - name: Check Removal Idempotency
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        vmid: "{{ create_res.vmid }}"
        state: absent
      retries: 1
      delay: 10
      register: check_res
      failed_when: check_res.changed

    - name: Create new LXC [With storage + disk] (should fail)
      community.general.proxmox:
        api_host: "{{ proxmox_api_host }}"
        api_user: "{{ proxmox_api_user }}"
        api_token_id: "{{ proxmox_api_token_id }}"
        api_token_secret: "{{ proxmox_api_token_secret }}"
        node: node01
        hostname: test-node-2
        ostemplate: local:vztmpl/debian-12-standard_12.2-1_amd64.tar.zst
        cores: 1
        disk: 6
        storage: local-lvm
      ignore_errors: true
```

### CODE REVIEW

File: plugins/modules/proxmox.py [hunk 1]
```diff
@@ -1067,257 +696,928 @@ def main():
...
-            module.fail_json(vmid=vmid, msg="deletion of VM %s failed with exception: %s" % (vmid, to_native(e)))
+
+def main():
+    module = get_ansible_module()
+    proxmox = ProxmoxLxcAnsible(module)
+
+    try:
+        proxmox.run()
+    except Exception as e:
+        module.fail_json(msg="An error occurred: %s" % to_native(e))
```
Comment (russoz) [inline plugins/modules/proxmox.py:1620]: You might want to consider using `ModuleHelper`. See more in https://docs.ansible.com/ansible/latest/collections/community/general/docsite/guide_modulehelper.html

Particularly for this module, I would recommend using `StateModuleHelper`, as it is already tailored for modules with different behaviour for each `state`. 

There is a number of modules already using these helpers, you can spot them using 
```bash
git grep -E '\(\S*ModuleHelper\)'
```

**Review by russoz — APPROVED** [review]
Hi @Lithimlin thanks for your contribution!

I would recommend breaking this PR into smaller pieces, as it seems to contain both bugfixes as well as new features. Additionally, we must not use type hints, as Ansible still provides support for Python 2.7 -  it's quite a nuisance, but that should be the case until that support is dropped, I want to say end of next year but I am not sure right now.

**Comment by bcoca** [comment]
>  Ansible still provides support for Python 2.7 - it's quite a nuisance, but that should be the case until that support is dropped, I want to say end of next year but I am not sure right now.

The last 2 versions of core already dropped support for 2.7 and 2.16 will be EOL in May 2025
https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#ansible-core-support-matrix

**Comment by russoz** [comment]
> > Ansible still provides support for Python 2.7 - it's quite a nuisance, but that should be the case until that support is dropped, I want to say end of next year but I am not sure right now.
> 
> The last 2 versions of core already dropped support for 2.7 and 2.16 will be EOL in May 2025 [docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#ansible-core-support-matrix](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#ansible-core-support-matrix)

Yes, but community.general still supports ansible-core 2.15. Felix mentioned the deadline more than once but it will be late next year - far enough into the future to be forgotten. Until then, in the collection, we must support Python 2.7.

File: plugins/modules/proxmox.py [hunk 2]
```diff
@@ -69,14 +69,14 @@
...
         type: str
       size:
         description:
           - O(disk_volume.size) is the size of the storage to use.
-          - The size is given in GB.
-          - Required only if O(disk_volume.storage) is defined and mutually exclusive with O(disk_volume.host_path).
-        type: int
+          - If no unit is added, the size is given in GB.
+          - Required only if O(disk_volume.storage) is defined, and mutually exclusive with O(disk_volume.host_path).
+        type: str
```
[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:80]: This new feature turned out to be more complex after all, so I removed it from the PR

File: plugins/modules/proxmox.py [hunk 3]
```diff
@@ -44,12 +44,12 @@
       - This option was previously described as "hard disk size in GB for instance" however several formats describing
         a lxc mount are permitted.
       - Older versions of Proxmox will accept a numeric value for size using the O(storage) parameter to automatically
-        choose which storage to allocate from, however new versions enforce the C(<STORAGE>:<SIZE>) syntax.
+        choose which storage to allocate from, however new versions enforce the C(<STORAGE>:<SIZE_IN_GiB>) syntax.
```
[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:48]: This is just a clarification taken from the proxmox docs

File: changelogs/fragments/9225-proxmox-module-refactoring.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+minor_changes:
+  - proxmox - refactors the proxmox module
```
[the remark this answers is not in the document]
Comment (russoz) [reply changelogs/fragments/9225-proxmox-module-refactoring.yml]: This was not solved

File: changelogs/fragments/9225-proxmox-module-refactoring.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+minor_changes:
+  - proxmox - refactors the proxmox module (https://github.com/ansible-collections/community.general/pull/9225)
+  - proxmox - fixes idempotency of template conversions (https://github.com/ansible-collections/community.general/pull/9225,https://github.com/ansible-collections/community.general/issues/8811).
+  - proxmox - fixes issues with disk_volume variable (https://github.com/ansible-collections/community.general/pull/9225,https://github.com/ansible-collections/community.general/issues/9065).
+  - proxmox - fixes incorrect parsing for bind-only mounts (https://github.com/ansible-collections/community.general/pull/9225,https://github.com/ansible-collections/community.general/issues/8982).
+  - proxmox module utils - fixes ignoring of ``choose_first_if_multiple`` argument in ``get_vmid``. (https://github.com/ansible-collections/community.general/pull/9225)
```
Comment (russoz) [inline changelogs/fragments/9225-proxmox-module-refactoring.yml:6]: Hi @Lithimlin sorry for being so picky about this, but thee lines are later published as the changelogs for the version - you can see what they will turn into at:
https://docs.ansible.com/ansible/latest/collections/community/general/changelog.html

So we have these "editorial" standards that we follow. In these lines:
```suggestion
  - proxmox - fixes idempotency of template conversions (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8811).
  - proxmox - fixes issues with disk_volume variable (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/9065).
  - proxmox - fixes incorrect parsing for bind-only mounts (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8982).
  - proxmox module utils - fixes ignoring of ``choose_first_if_multiple`` argument in ``get_vmid`` (https://github.com/ansible-collections/community.general/pull/9225).
```

File: changelogs/fragments/9225-proxmox-module-refactoring.yml [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+deprecated_features:
+  - proxmox - removes default value of ``update`` parameter. This will be changed to a default of ``true`` in the future.
```
Comment (russoz) [inline changelogs/fragments/9225-proxmox-module-refactoring.yml:9]: Missing the PR link here as well

**Review by russoz — CHANGES_REQUESTED** [review]
Hi, I had approved it before, but will mark it "Request changes" to prevent other maintainers from accidentally merging it as-is now. Also, on second thought, this is quite a big revamp, so it would benefit from a more thorough examination before we merge it. I approved it before (and I should not, because the changelog needed fixing) and, days later, I cannot recall what I have seen here or not. I'm writing this to manage the expectations - this PR might take a tad longer to be merged - but please bear with us.

**Comment by Lithimlin** [comment]
> this is quite a big revamp, so it would benefit from a more thorough examination before we merge it.

Yes, that would be great. While I have tested quite a few cases, I can't test all corner cases that might occur in my current setup. Having more people take a look at this and run some tests would be great.

I will also take a look at the `[State]ModuleHelper`, but I find the approach of "documentation by example" a bit hard to work with.

**Comment by russoz** [comment]
> > this is quite a big revamp, so it would benefit from a more thorough examination before we merge it.
> 
> Yes, that would be great. While I have tested quite a few cases, I can't test all corner cases that might occur in my current setup. Having more people take a look at this and run some tests would be great.

Sweet. Thanks for your patience.

> I will also take a look at the `[State]ModuleHelper`, but I find the approach of "documentation by example" a bit hard to work with.

There is proper documentation at: https://docs.ansible.com/ansible/latest/collections/community/general/docsite/guide_modulehelper.html

(unless that's what you are referring to, in which case I would gladly take suggestions on how to improve those docs)

**Comment by felixfontein** [comment]
> > > Ansible still provides support for Python 2.7 - it's quite a nuisance, but that should be the case until that support is dropped, I want to say end of next year but I am not sure right now.
> > 
> > 
> > The last 2 versions of core already dropped support for 2.7 and 2.16 will be EOL in May 2025 [docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#ansible-core-support-matrix](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#ansible-core-support-matrix)
> 
> Yes, but community.general still supports ansible-core 2.15. Felix mentioned the deadline more than once but it will be late next year - far enough into the future to be forgotten. Until then, in the collection, we must support Python 2.7.

We drop support for one ansible-core version per major release (unless we forget to do it, like for 9.0.0, and then we dropped two in 10.0.0 ;) ). We support all ansible-core versions that have not been EOL for more than a short time when we release a new major version. Since ansible-core 2.16 becomes EOL in May 2025, around when we have a new major release (11.0.0), we'll only drop support for it in the next major release (12.0.0, in ~November 2025). Only from that point on Python 2.7 is no longer supported by community.general.

**Comment by Lithimlin** [comment]
I've looked at a re-design using the `StateModuleHelper`, and while that would be very convenient indeed, I think it would require a rework of the `module_utils.proxmox` if I understood correctly since `ProxmoxAnsible` uses the base `AnsibleModule` in its `self.module` quite a lot. This, in turn, would require a re-design of all other `proxmox_*` modules, which is probably a good idea but for which I just do not have the time and resources.

I would like to change the implementation to one using the `StateModuleHelper` and take the first step towards refactoring all the modules, but with the additional overhead from above, it is simply not feasible for me to do so right now.

File: plugins/modules/proxmox.py (hunk 1 above)
Comment (Lithimlin) [reply plugins/modules/proxmox.py:1728]: see [this comment](https://github.com/ansible-collections/community.general/pull/9225#issuecomment-2548649397) for a reply

**Comment by felixfontein** [comment]
Please note that there is now a conflict (after we reformatted most module documentations).

To rebase, it might be easier to first squash all commits into one, and *then* rebase against the current `main`. Otherwise you might have to resolve conflicts several times.

**Comment by felixfontein** [comment]
It looks like there's anohter conflict now (I guess due to #9524). Sorry for the hassle :( I hope this was the last reformat...

File: changelogs/fragments/9225-proxmox-module-refactoring.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+minor_changes:
+  - proxmox - refactors the proxmox module (https://github.com/ansible-collections/community.general/pull/9225)
+  - proxmox - fixes idempotency of template conversions (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8811).
+  - proxmox - fixes issues with disk_volume variable (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/9065).
+  - proxmox - fixes incorrect parsing for bind-only mounts (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8982).
+  - proxmox module utils - fixes ignoring of ``choose_first_if_multiple`` argument in ``get_vmid`` (https://github.com/ansible-collections/community.general/pull/9225)
```
Comment (felixfontein) [inline changelogs/fragments/9225-proxmox-module-refactoring.yml:6]:
```suggestion
  - proxmox module utils - fixes ignoring of ``choose_first_if_multiple`` argument in ``get_vmid`` (https://github.com/ansible-collections/community.general/pull/9225).
```

File: changelogs/fragments/9225-proxmox-module-refactoring.yml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+minor_changes:
+  - proxmox - refactors the proxmox module (https://github.com/ansible-collections/community.general/pull/9225)
+  - proxmox - fixes idempotency of template conversions (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8811).
```
Comment (felixfontein) [inline changelogs/fragments/9225-proxmox-module-refactoring.yml:3]: Shouldn't tihs and the next fragments be moved to `bugfixes`?
```suggestion

bugfixes:
  - proxmox - fixes idempotency of template conversions (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8811).
```

File: changelogs/fragments/9225-proxmox-module-refactoring.yml [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+  - proxmox - fixes issues with disk_volume variable (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/9065).
+  - proxmox - fixes incorrect parsing for bind-only mounts (https://github.com/ansible-collections/community.general/pull/9225, https://github.com/ansible-collections/community.general/issues/8982).
+  - proxmox module utils - fixes ignoring of ``choose_first_if_multiple`` argument in ``get_vmid`` (https://github.com/ansible-collections/community.general/pull/9225)
+
+deprecated_features:
+  - proxmox - removes default value of ``update`` parameter. This will be changed to a default of ``true`` in the future (https://github.com/ansible-collections/community.general/pull/9225)
```
Comment (felixfontein) [inline changelogs/fragments/9225-proxmox-module-refactoring.yml:9]:
```suggestion
  - proxmox - deprecates default value ``false`` of ``update`` parameter. This will be changed to a default of ``true`` in community.general 11.0.0 (https://github.com/ansible-collections/community.general/pull/9225).
```

File: plugins/modules/proxmox.py [hunk 10]
```diff
@@ -1044,282 +676,1059 @@ def main():
...
-            if proxmox.start_instance(vm, vmid, timeout):
-                module.exit_json(changed=True, vmid=vmid, msg="VM %s started" % vmid)
-        except Exception as e:
-            module.fail_json(vmid=vmid, msg="starting of VM %s failed with exception: %s" % (vmid, e))
+        # check if the container exists already
+        if lxc is not None:
+            if update is None:
+                # TODO: Remove deprecation warning in version 11.0.0
+                self.module.deprecate(
+                    msg="The default value of False for 'update' has been deprecated and will be changed to True in version 11.0.0.",
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:853]: YAML booleans are lowercase:
```suggestion
                    msg="The default value of false for 'update' has been deprecated and will be changed to true in version 11.0.0.",
```

File: plugins/modules/proxmox.py [hunk 11]
```diff
@@ -186,7 +188,7 @@
   storage:
     description:
       - Target storage.
-      - This Option is mutually exclusive with O(disk) and O(disk_volume).
+      - This Option is mutually exclusive with O(disk_volume) and O(mount_volumes).
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:191]:
```suggestion
      - This option is mutually exclusive with O(disk_volume) and O(mount_volumes).
```

File: plugins/modules/proxmox.py [hunk 12]
```diff
@@ -226,8 +228,8 @@
   update:
     description:
       - If V(true), the container will be updated with new values.
+      - The current default value of V(false) is deprecated and should be replaced with V(update=true) in version 11.0.0.
```
Comment (felixfontein) [inline plugins/modules/proxmox.py:231]:
```suggestion
      - The current default value of V(false) is deprecated and will change to V(true) in community.general 11.0.0.
        Please set O(update) explicitly to V(false) or V(true) to avoid surprises and get rid of the deprecation warning.
```

**Review by russoz — COMMENTED** [review]
Hi @Lithimlin 

I have spent more time looking into it and made it through `state=present`. Got some comments on it. Cheers

File: plugins/modules/proxmox.py [hunk 13]
```diff
@@ -1044,282 +676,1059 @@ def main():
...
+                    nameserver=self.params.get("nameserver"),
+                    netif=self.params.get("netif"),
+                    onboot=ansible_to_proxmox_bool(self.params.get("onboot")),
+                    searchdomain=self.params.get("searchdomain"),
+                    startup=startup,
+                    swap=self.params.get("swap"),
+                    tags=self.params.get("tags"),
+                    timezone=self.params.get("timezone"),
+                )
+                self.module.exit_json(
```
[the remark this answers is not in the document]
Comment (russoz) [reply plugins/modules/proxmox.py:893]: Please read the next one below before doing anything here.

File: plugins/modules/proxmox.py [hunk 14]
```diff
@@ -1044,282 +676,1059 @@ def main():
...
+            self.lxc_to_template(
+                vmid,
+                hostname,
+                node=self.params.get("node"),
+                timeout=self.params.get("timeout"),
+                force=self.params.get("force"),
+            )
+
+    def lxc_present(self, vmid, hostname, node, update, force):
+        self.check_supported_features()
```
[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:833]: I thought this was only relevant for `state="present"`, but it probably also applies to `template`. I'll move this up a bit.

File: plugins/modules/proxmox.py [hunk 15]
```diff
@@ -1044,282 +676,1059 @@ def main():
...
+                ["{0}={1}".format(k, v) for k, v in kwargs.items()]
+            )
+        return {key: vol_string}
 
-    elif state == 'restarted':
-        try:
-            vm = proxmox.get_vm(vmid)
+    def get_lxc_resource(self, vmid, hostname):
+        if not vmid and not hostname:
+            self.module.fail_json(msg="Either VMID or hostname must be provided.")
```
[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:1600]: I can add the `required_one_of`, but this is more a check in place for the function itself so it is not misused.

File: plugins/modules/proxmox.py [hunk 16]
```diff
@@ -1044,282 +676,1059 @@ def main():
...
 
-            if (proxmox.stop_instance(vm, vmid, timeout, force=module.params['force']) and
-                    proxmox.start_instance(vm, vmid, timeout)):
-                module.exit_json(changed=True, vmid=vmid, msg="VM %s is restarted" % vmid)
+        vms = [vm for vm in vms if vm["vmid"] == vmid]
+        if len(vms) == 0:
+            raise LookupError("VM with VMID %d does not exist in cluster." % vmid)
+
+        return vms[0]
+
```
[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:1624]: Somewhere down the line I'd love to put `get_vm_resources()` and `get_lxc_resource_by_id()` into the utils module, but I didn't want to replace it there as that would require me to check all other modules for functionality as well. I have neither the setup nor the time to do that unfortunately.

What I could do is add both of these new functions to the utils module so they can be used down the line and possibly deprecate the old one, but I'm not sure how good of an idea that is.

File: plugins/modules/proxmox.py [hunk 17]
```diff
@@ -1044,282 +676,1059 @@ def main():
...
+    def lxc_present(self, vmid, hostname, node, update, force):
+        self.check_supported_features()
 
-    elif state == 'started':
         try:
-            vm = proxmox.get_vm(vmid)
-            if getattr(proxmox.proxmox_api.nodes(vm['node']), VZ_TYPE)(vmid).status.current.get()['status'] == 'running':
-                module.exit_json(changed=False, vmid=vmid, msg="VM %s is already running" % vmid)
+            lxc = self.get_lxc_resource(vmid, hostname)
```
[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:836]: While that is true, this ties into the discussion of dropping `openvz` support since it's from a Proxmox version that is way past its EOL.

File: plugins/modules/proxmox.py [hunk 18]
```diff
@@ -1044,282 +675,1055 @@ def main():
...
+                changed=True,
+                vmid=vmid,
+                msg="Created VM %s from template %s" % (identifier, ostemplate),
+            )
+
+        self.module.fail_json(
+            vmid=vmid,
+            msg="VM %s does not exist but neither clone nor ostemplate were specified!"
+            % identifier,
+        )
```
Comment (russoz) [inline plugins/modules/proxmox.py:1167]: Is this case ever going to happen? Line 742 (required_if state present) guarantee at least one of `update` (in which case this method `new_lxc_instance()` does not get called), `clone` or `ostemplate` is passed. Please double check, if sure it won't happen, please remove.

Comment (Lithimlin) [reply plugins/modules/proxmox.py:1167]: It is possible for the LXC to not exist, but `update` being set to `true`, in which case neither `clone` nor `ostemplate` need to be set according to the module definition.

File: plugins/modules/proxmox.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (russoz) [reply plugins/modules/proxmox.py:1595]: The functions within this file are not meant to be imported elsewhere, so in theory if not in practice, the only misuse that _can_ happen is in this file as well. If you could guarantee that is not the case, then this check could be removed.

[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:1595]: I suppose that means I should properly document all functions in this module as well

[the remark this answers is not in the document]
Comment (Lithimlin) [reply plugins/modules/proxmox.py:1595]: Ah, actually, I remember why I did it this way: This is preparation for being able to safely use this function in the `module_utils`.

**Comment by felixfontein** [comment]
@russoz @Lithimlin what's the current state of the PR? Is everyone happy so we can merge this for 10.3.0? I guess we won't get more feedback from users of the module here, unless it ends up in a release...

**Review by russoz — APPROVED** [review]
Well, I'm approving on the basis that it seems to have been carefully tailored, but I have not really examined all the logic paths. Again, this PR is way too big - it would have been better to create a number of smaller ones.
@Lithimlin I understand/assume you have tested this thoroughly and I am approving based on that.

**Comment by felixfontein** [comment]
@Lithimlin thanks a lot for your contribution!
@russoz @bcoca thanks a lot for reviewing this!
